diff --git a/patch_cirrus/patch_cirrus_apple.h b/patch_cirrus/patch_cirrus_apple.h
index 398618c..0abda63 100644
--- a/patch_cirrus/patch_cirrus_apple.h
+++ b/patch_cirrus/patch_cirrus_apple.h
@@ -157,7 +157,6 @@ struct sub_codec cs8409_cs42l83_codec = {
         .hp_jack_in = 0,
         .mic_jack_in = 0,
         .linein_jack_in = 0,
-        .force_status_change = 1,
         .paged = 1,
         .suspended = 1,
         .no_type_dect = 0,
